executor/analyze.go

```diff
@@ -496,22 +527,458 @@ type AnalyzeFastTask struct {
 // AnalyzeFastExec represents Fast Analyze executor.
 type AnalyzeFastExec struct {
 	ctx             sessionctx.Context
-	PhysicalTableID int64
+	physicalTableID int64
 	pkInfo          *model.ColumnInfo
 	colsInfo        []*model.ColumnInfo
 	idxsInfo        []*model.IndexInfo
 	concurrency     int
 	maxNumBuckets   uint64
-	table           table.Table
 	cache           *tikv.RegionCache
 	wg              *sync.WaitGroup
 	sampLocs        chan *tikv.KeyLocation
-	sampLocRowCount uint64
-	tasks           chan *AnalyzeFastTask
+	rowCount        uint64
+	sampCursor      int32
+	sampTasks       []*AnalyzeFastTask
 	scanTasks       []*tikv.KeyLocation
+	collectors      []*statistics.SampleCollector
+}
+
+func (e *AnalyzeFastExec) getSampRegionsRowCount(bo *tikv.Backoffer, needRebuild *bool, err *error, sampTasks *[]*AnalyzeFastTask) {
+	defer func() {
+		e.wg.Done()
+		if *needRebuild == true {
+			close(e.sampLocs)
```

Comment:
`getSampRegionsRowCount()` is the consumer of this channel, note that, this channel is also closed in the producer side, which is `buildSampTask`. Please do not close it here to avoid panic caused by closing a closed channel.